### PR TITLE
RATIS-2066. RaftServerProxy supports close callback

### PR DESCRIPTION
**Pull request by ChenSammi**
## What changes were proposed in this pull request?
Currently there is no way for application to get notified once the RaftServerProxy is close.

This task is to add a close callback support to RaftServerProxy. 

Motivated by HDDS-10749. 

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/RATIS-2066

## How was this patch tested?
integration test with Ozone.

### CODE REVIEW

**Comment by szetszwo** [comment]
@ChenSammi  , thanks for working on this!  

We already have a `notifyServerShutdown`.  It will notify the `StateMachine` when its division (i.e. `RaftServerImpl`, not `RaftServerProxy`) is closed.  Is it good enough?
https://github.com/apache/ratis/blob/1fe59e07dff9b02e46d150de8b46ed2d4fc3abce/ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java#L268

**Comment by ChenSammi** [comment]
@szetszwo , notifyServerShutdown is not enough for this DN case. For DN, there is always new pipeline created and old pipeline destroyed.  I guess this notifyServerShutdown will be called when an old pipeline is closed.  This specific DN case, is when there is long GC, longer than configured threshold, RaftServerProxy will close itself. When this happens, we need to close DN as well. To achieve this, one solution is provide an close callback as this JIRA do, another solution is we start a new thread in DN to closely monitor the RaftServerProxy state, whether it's running or closed.  After some thoughts, I prefer the first solution since it will save a thread resource.

**Comment by szetszwo** [comment]
@ChenSammi , we may add a boolean parameter as below.  Anyway, we should use the same mechanism to notify the `StateMachine`.
```java
    /**
     * Notify the {@link StateMachine} that a raft server has step down.
     */
    default void notifyServerShutdown(RoleInfoProto roleInfo) {
      notifyServerShutdown(roleInfo, false);
    }

    default void notifyServerShutdown(RoleInfoProto roleInfo, boolean allServers) {}
```

**Comment by ChenSammi** [comment]
@szetszwo , notifyServerShutdown is called currently when a peer's in vote phase, and is found not in the raft group by leader, so it turns down itself.  

The case we met in Ozone datanode is different.  The trace is,
  RaftServerProxy#pauseMonitor#handleJvmPause -> RaftServerProxy#close -> ImplMap for each Impl, RaftServerImpl#close

None of StateMachine#EventApi is called now.  Are you suggesting to call notifyServerShutdown in RaftServerImpl#close, and give a reason why RaftServerImpl shutdown itself?

**Comment by ChenSammi** [comment]
@szetszwo ,  could you re-trigger the CI?  The failed TestSegmentedRaftLog passed locally.

File: ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java [hunk 1]
```diff
@@ -266,6 +266,13 @@ default void notifySnapshotInstalled(InstallSnapshotResult result, long snapshot
      * Notify the {@link StateMachine} that a raft server has step down.
      */
     default void notifyServerShutdown(RoleInfoProto roleInfo) {}
```
Comment (szetszwo) [inline ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java:268]: We should deprecate the old method and change its default to
```java
    /**
     * The same as notifyServerShutdown(roleInfo, false).
     *
     * @deprecated Implementations should override {@link #notifyServerShutdown(RoleInfoProto, boolean)} instead.
     */
    @Deprecated
    default void notifyServerShutdown(RoleInfoProto roleInfo) {
      notifyServerShutdown(roleInfo, false);
    }
```
Could you also update the caller?
```java
+++ b/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java
@@ -334,7 +334,7 @@ class LeaderElection implements Runnable {
         case NOT_IN_CONF:
         case SHUTDOWN:
           server.close();
-          server.getStateMachine().event().notifyServerShutdown(server.getRoleInfoProto());
+          server.getStateMachine().event().notifyServerShutdown(server.getRoleInfoProto(), false);
           return false;
         case TIMEOUT:
```

File: ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java [hunk 2]
```diff
@@ -266,6 +266,13 @@ default void notifySnapshotInstalled(InstallSnapshotResult result, long snapshot
...
+
+    /**
+     * Notify the {@link StateMachine} that a raft server has been shutdown down.
```
Comment (szetszwo) [inline ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java:271]: `a raft server` seems inaccurate.  How about the following ?
```java
     * Notify the {@link StateMachine} that either the server for this division or all the servers have been shut down.
```

**Review by szetszwo — COMMENTED** [review]
@ChenSammi , thanks a lot for the update!  Please see the comments inlined.

File: ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java (hunk 1 above)
Comment (ChenSammi) [reply ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java]: Sure. Addressed in new commit.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by ChenSammi** [comment]
Thanks @szetszwo .

**Pull request merged**